src/PerformanceSensitiveAnalyzers/UnitTests/TypeConversionAllocationAnalyzerTests.cs

```diff
@@ -686,5 +692,120 @@ public void SomeMethod()
                 VerifyCS.Diagnostic(TypeConversionAllocationAnalyzer.DelegateOnStructInstanceRule).WithLocation(12, 22));
         }
+
+
+        [Fact]
+        public async Task TypeConversionAllocation_NoDiagnosticWhenPassingDelegateAsArgument()
+        {
+            const string snippet = @"
+using System;
+using Roslyn.Utilities;
+
+struct Foo
+{
+    [PerformanceSensitive(""uri"")]
+    void Do(Action process)
+    {
+        DoMore(process);
+    }
+
+    void DoMore(Action process)
+    {
+        process();
+    }
+}
+            ";
+            await VerifyCS.VerifyAnalyzerAsync(snippet);
+        }
+
+        [Fact]
+        public async Task TypeConversionAllocation_ReportBoxingAllocationForPassingStructInstanceMethodForDelegateConstructor()
+        {
+            const string snippet = @"
+using System;
+using Roslyn.Utilities;
+
+public struct MyStruct
+{
+    [PerformanceSensitive(""uri"")]
+    public void Testing()
+    {
+        var @struct = new MyStruct();
+        @struct.ProcessFunc(new Func<object, string>(FooObjCall));
+    }
+
+    public void ProcessFunc(Func<object, string> func)
+    {
+    }
+
+    private string FooObjCall(object obj)
+    {
+        return obj.ToString();
+    }
+}
+            ";
+
+            await VerifyCS.VerifyAnalyzerAsync(snippet,
+                // Test0.cs(11,54): warning HAA0602: Struct instance method being used for delegate creation, this will result in a boxing instruction
+                VerifyCS.Diagnostic(TypeConversionAllocationAnalyzer.DelegateOnStructInstanceRule).WithSpan(11, 54, 11, 64));
+        }
+
+        [Fact]
+        public async Task TypeConversionAllocation_DoNotReportBoxingAllocationForPassingStructStaticMethodForDelegateConstructor()
+        {
+            const string snippet = @"
+using System;
+using Roslyn.Utilities;
+
+public struct MyStruct
+{
+    [PerformanceSensitive(""uri"")]
+    public void Testing()
+    {
+        var @struct = new MyStruct();
+        @struct.ProcessFunc(new Func<object, string>(FooObjCall));
+    }
+
+    public void ProcessFunc(Func<object, string> func)
+    {
+    }
+
+    private static string FooObjCall(object obj)
+    {
+        return obj.ToString();
+    }
+}
+            ";
+
+            await VerifyCS.VerifyAnalyzerAsync(snippet);
+        }
+
+        [Fact]
+        public async Task TypeConversionAllocation_DoNotReportInlineDelegateAsStructInstanceMethods()
+        {
+            const string snippet = @"
+using System;
+using Roslyn.Utilities;
+
+public struct MyStruct
+{
+    [PerformanceSensitive(""uri"")]
+    public void Testing()
+    {
+        var ints = new[] { 5, 4, 3, 2, 1 };
+        Array.Sort(ints, delegate(int x, int y) { return x - y; });
+        Array.Sort(ints, (x, y) => x - y);
+        DoSomething(() => throw new Exception());
+        DoSomething(delegate() { throw new Exception(); });
```

Comment:
💡 This needs a test with the simple lambda syntax as well (`x => throw new Exception()`)